server/src-lib/Hasura/GraphQL/Transport/WebSocket.hs

```diff
@@ -265,154 +266,206 @@ onStart serverEnv wsConn (StartMsg opId q) =
+    execPlans <-
+      either (withComplete . preExecErr requestId) (return . toList) execPlanE
+    let execCtx =
+          E.ExecutionCtx
+            logger
+            sqlGenCtx
+            pgExecCtx
+            planCache
+            sc
+            scVer
+            httpMgr
+            enableAL
+    case getSubscriptionPlan execPlans
+      -- When not a subscription operation
+          of
+      Nothing -> do
+        logOpEv ODStarted (Just requestId)
+        results <-
           liftIO $
-            STM.atomically $ STMMap.insert (lqId, _grOperationName q) opId opMap
-          logOpEv ODStarted (Just reqId)
-
+          runExceptT $ do
+            flip mapM execPlans $ \execPlan ->
+              case execPlan of
+                E.Leaf plan ->
+                  case plan of
+                    E.ExPHasura op ->
+                      case op of
+                        E.ExOpQuery opTx genSql ->
+                          fmap (\res -> encJFromAssocList [("data", res)]) $
+                          execQueryOrMut requestId q genSql $
+                          runLazyTx' pgExecCtx opTx
+                        E.ExOpMutation opTx ->
+                          fmap (\res -> encJFromAssocList [("data", res)]) $
+                          execQueryOrMut requestId q Nothing $
+                          runLazyTx pgExecCtx $ withUserInfo userInfo opTx
+                        E.ExOpSubs {} ->
+                          throwError
+                            (err500
+                               Unexpected
+                               "did not expect subscription operation here")
+                    E.ExPRemote rtf ->
+                      runRemoteGQ execCtx requestId userInfo reqHdrs rtf
+                E.Tree {} ->
+                  throwError
+                    (err400
+                       NotSupported
+                       "remote relationships not supported over websocket")
+        case results of
+          Left err -> postExecErr requestId err
+          Right results' -> do
+            let mergedResponse = HTTP.mergeResponseData results' True
+            case mergedResponse of
+              Left e ->
+                postExecErr requestId $
+                err500
+                  UnexpectedPayload
+                  ("could not merge data from results: " <> T.pack e)
+              Right resp -> do
+                sendSuccResp resp
+                sendCompleted (Just requestId)
+      Just subPlans ->
+        case subPlans of
+          [] -> preExecErr requestId (err500 Unexpected "subscription has no plan")
+          [x] -> do
+            logOpEv ODStarted (Just requestId)
+            execSubscription requestId q x
+          _ -> preExecErr requestId (err500 Unexpected "only one subscription plan expected")
+    return ()
+  where
+    getSubscriptionPlan plans =
+      let subPlans = map (getSubPlan . getLeafPlan) plans
+       in case catMaybes subPlans of
```

Comment:
You should just do `... in catMaybes subPlans` I think. By returning `Maybe [a]` you introduce an unreachable error case you have to handle on line 344 (IIUC)

Reply:
In that case it will return `[a]` which will be worse to pattern match no?

Currently, I am pattern matching over "no subscription plan" or "subscription plans". It would have been nicer if I could just catch the error in `getSubscriptionPlan` but the monadic constraints are so different that I am unable to use `preExecError` inside `getSubscriptionPlan`.

It is possible that I haven't understood your comment 100%, let me know if that's the case.